Feature/video thumbnail support

bastiion opened the pull request:
this includes the ability to create thumbnails of videos, by using libav (ffmpeg-next) ffmpeg, mpv. I hope it fits into the pipline.

bastiion committed:
use env var for data folder

bastiion committed:
Refactor cache path to use env::temp_dir()

bastiion committed:
Add support for generating video thumbnails using mpv

bastiion committed:
docs: update README to include Nix flake instructions and supported video formats

lamka02sk commented:
Hello,
although very interesting and handy, unfortunately, it doesn't seem like this feature is ready for merge and we need to look at it from a little bit different perspective to make it work with picturium.

**1. Easy to use**
picturium aims to be easy to use and that is the main reason why I opted for docker container as the main and recommended way to run picturium. Providing `mpv` executable through environment variable makes it quiet hard with docker. If there are any dynamically linked libraries in `mpv` executable (which there probably are), it will fail inside docker container and will be a pain to find all the files required and then add them as volume, or worse building custom docker image with mpv installed. 
This approach can also cause unforseen problems when using different versions of `mpv`.
picturium should use system `mpv` executable included in `$PATH` and with that, the executable should be a part of the provided docker image. Also this way, there is only one officially supported version of `mpv` that can be tested and guaranteed to be working.
I think dropping the `MPV` environment variable and rather letting admins decide the way of providing `mpv` executable would be a better approach in 99% of cases.

**2. Works with picturium parameters**
You seem to be using `width` URL parameter to determine thumbnail size, but for example `height` is ignored. Also when neither `width` or `height` is present, the output size should be the original size. Thumbnail dimensions should be determined similarly to function used while rasterizing vector images before processing: https://github.com/lamka02sk/picturium/blob/master/src/pipeline/resize.rs#L89

**3. Finding the best thumbnail**
The current code seems to be always selecting the first frame of video since the thumbnail file size is being compared with 0 during the first iteration. I haven't actually tested it, so I am sorry, if it actually works, maybe I missed something.

**4. Performance**
Another aim of picturium is to be fast. Have you tried different methods of generating video thumbnails, for example using `ffmpeg`? With some limited testing I have done on a 1 GB H.265 file, `ffmpeg` seemed to yield about 33% better performance, but it might be an isolated case.

bastiion commented:
Nice to see this in-depth review, I hope to find some time soon to test you propositions and make it into the code.
the addition with the PATH instead of directly giving the MPV path is also a good advice. Would you prefer, to include both options -> MPV and FFMPEG, so that user can decide which one to use? It could be that MPV supports even more formats then FFMPeg, but this is just a quick guess.

Thank you for the effort you already did in implementing this thumbnail service.

lamka02sk commented:
> Would you prefer, to include both options -> MPV and FFMPEG, so that user can decide which one to use? It could be that MPV supports even more formats then FFMPeg, but this is just a quick guess.

Yeah, that would be awesome to implement both mpv and ffmpeg backends for this feature 👍🏼

bastiion committed:
Add dimension calculation and cache directory helper functions

bastiion committed:
Refactor video thumbnail generation with ffmpeg and mpv backends

- Add generate_thumbnail_mpv() and generate_thumbnail_ffmpeg() functions
- Implement proper dimension calculation respecting both width and height parameters
- Fix best thumbnail selection logic to properly compare file sizes
- Support configurable thumbnail positions via VIDEO_THUMBNAIL_POSITIONS
- Add main dispatcher that selects backend based on VIDEO_BACKEND env var
- Use system executables from PATH instead of MPV env var

bastiion committed:
Rename mpv_thumb module to video_thumb

Reflects support for multiple video thumbnail backends (ffmpeg and mpv)

bastiion committed:
docs: update README for video thumbnail multi-backend support

- Document VIDEO_BACKEND and VIDEO_THUMBNAIL_POSITIONS environment variables
- Add configuration examples for ffmpeg and mpv backends
- Update supported formats list to mention both ffmpeg and mpv

bastiion committed:
Add optional native-ffmpeg feature with ffmpeg-next dependency

- Add native-ffmpeg feature flag as opt-in
- Add ffmpeg-next 8.0.0 as optional dependency
- Maintains backward compatibility (default build unchanged)

bastiion committed:
Add NativeFFmpeg variant to VideoBackend enum

- Conditionally compiled with native-ffmpeg feature
- No impact on default build

bastiion committed:
Update backend detection to prioritize native-ffmpeg

- detect_available_backend() returns NativeFFmpeg when feature enabled
- get_video_backend() supports VIDEO_BACKEND=native
- Auto mode priority: native > ffmpeg > mpv

bastiion committed:
Implement native ffmpeg backend with in-process decoding

- Add generate_thumbnail_native_ffmpeg() function with full video decoding
- Support percentage and frame-based seeking with duration calculation
- Implement frame scaling using ffmpeg's software scaler
- Save frames as PNG via PPM intermediate format
- Integrate into main dispatcher with conditional compilation
- Requires native-ffmpeg feature flag to compile

bastiion committed:
docs: update README with native ffmpeg backend information

- Document all three video thumbnail backends (native, ffmpeg, mpv)
- Add build instructions for native-ffmpeg feature
- Explain backend priority and performance characteristics
- Update examples to include native backend usage

bastiion committed:
Add helper functions for video backend detection and position parsing

bastiion committed:
Fix native ffmpeg aspect ratio handling for non-square pixels

- Calculate DAR from SAR for proper aspect ratio
- Handle frame stride to avoid visual artifacts
- Scale to display dimensions for original size requests

bastiion committed:
docs: remove Nix Flake section from README

bastiion commented:
Hi @lamka02sk ,

Thank you for the detailed feedback. I've gone through all your points and refactored the video thumbnail generation accordingly. Here's what changed:

**1. Easy to use - System executables**
Dropped the MPV environment variable completely. The system now uses executables from PATH (ffmpeg and mpv), which makes it work seamlessly in Docker containers. Just install the backend you want, and it will be detected automatically.

**2. Parameter handling**
Fixed the dimension calculation to properly handle both width and height parameters. The logic now works similar to the vector image resize function:
- No parameters: Uses original video dimensions
- Only width: Auto-calculates height maintaining aspect ratio
- Only height: Auto-calculates width maintaining aspect ratio  
- Both: Uses specified values

**3. Best thumbnail selection**
You were right about the bug. Fixed the file size comparison logic so it actually compares between frames instead of always picking the first one: `if file_size > 0 && file_size > best_size`

**4. Performance - Multiple backends**
Implemented both ffmpeg and mpv backends as you suggested. Added a VIDEO_BACKEND environment variable (auto/ffmpeg/mpv) that defaults to auto, which prefers ffmpeg first and falls back to mpv if needed.

**Additional improvements:**
- Made sampling positions configurable via VIDEO_THUMBNAIL_POSITIONS (supports percentages like "25%,50%" or frame numbers like "100,200")
- Added a third optional backend using ffmpeg_next with native libav bindings for in-process thumbnail generation, avoiding process spawning overhead entirely. This is behind a Cargo feature flag (native-ffmpeg) so it's completely optional.
- Fixed aspect ratio handling for videos with non-square pixels (SAR/DAR calculation)
- Proper stride handling to avoid visual artifacts in decoded frames

The native ffmpeg backend works really well for scenarios where you want maximum performance and have control over the build environment. The command-line backends (ffmpeg/mpv) remain the default and work without any special build requirements.

All changes are backwards compatible. The defaults work out of the box with system ffmpeg or mpv, and the advanced options are there for users who need them.

Let me know if there's anything else that needs adjustment.

bastiion committed:
fix: remove duplicate version line in Cargo.lock

bastiion committed:
cleanup: remove mpv_thumb.rs in favour of video_thumb.rs

lamka02sk commented on .env.example:
```diff
@@ -13,4 +13,6 @@ AVIF_ENABLE=true
\ No newline at end of file
+CACHE_CAPACITY=10
+
+DATA_DIR=./data
```
`DATA_DIR` default value should be an empty string or a dot.

lamka02sk commented on src/services/formats.rs:
```diff
@@ -102,7 +103,7 @@ pub fn determine_output_format(url_parameters: &UrlParameters, accept: Option<&H
 pub fn is_thumbnail_format(path: &Path) -> bool {
     let extension = get_extension(path).unwrap_or_else(|_| String::new());
-    matches!(extension.as_str(), "pdf" | "doc" | "docx" | "odt" | "xls" | "xlsx" | "ods" | "ppt" | "pptx" | "odp" | "rtf")
+    matches!(extension.as_str(), "pdf" | "doc" | "docx" | "odt" | "xls" | "xlsx" | "ods" | "ppt" | "pptx" | "odp" | "rtf" | "mp4" | "mkv" | "webm")
```
Some video formats are missing from this list. Because of this, the thumbnails are trying to be generated by libvips, which instantly fails.

lamka02sk reviewed (changes requested):
Hello @bastiion ,
sorry for the delay. I tested everything and found a few kinda minor bugs. Except the things I will mention, the video thumbnail generation worked very well and I am really happy with the feature.


## Data directory compatibility issue

The default value of `DATA_DIR` env variable is not backwards compatible with the latest picturium version. With the new proposed default value, URLs in all existing picturium instances will change and instances working with multiple directories will stop working altogether. The default should be empty string or a dot (I vote for empty string).


## Using CLI version of ffmpeg doesn't work

Default value of `VIDEO_THUMBNAIL_POSITIONS` env variable uses percentages to determine thumbnail generation points. The problem is that ffmpeg CLI does not support percentages as values of `-ss` option.

lamka02sk commented on src/services/mod.rs:
```diff
@@ -22,13 +22,16 @@ pub mod scheduler;
 pub async fn serve(req: HttpRequest, path: Path<String>, parameters: Query<HashMap<String, String>>, raw_url_parameters: Query<RawUrlParameters>) -> impl Responder {
 
-    let path = path.into_inner();
+    let data_dir = env::var("DATA_DIR").unwrap();
```
Using `unwrap()` is going to cause problems during update of existing picturium instances, since it is now required env variable. Instead, it should use `unwrap_or(DEFAULT_DATA_DIR)` to automatically set the default when .env doesn't contain `DATA_DIR` variable.

lamka02sk commented:
v0.2 will be coming out soon with video processing included

bastiion commented:
Okay, I see. You've basically re-implemented it. I'm happy to see it in the new release. And I think it's a good decision to simplify it to one backend to keep code maintenance low. 

I'm looking forward to the release. Thanks for your work.
